worksheet.rb

```diff
@@ -23,12 +38,209 @@
+            date: '5th Feb 2016',
+            cost: 20,
+            rider_id: 'RD0073',
+            rating: 5
+        },
+        {
+            driver_id: 'DR0001',
+            date: '5th Feb 2016',
+            cost: 45,
+            rider_id: 'RD0003',
+            rating: 2
+        }
+    ]
+}
+
 ########################################################
 # Step 4: Total Driver's Earnings and Number of Rides
 
 # Use an iteration blocks to print the following answers:
 # - the number of rides each driver has given
+def num_rides(structured_data)
+  driver_ride_count = {}
+  structured_data[:rides].each do |ride|
+    if driver_ride_count.include?(ride[:driver_id])
+      driver_ride_count[ride[:driver_id]] += 1
+    else
+      driver_ride_count[ride[:driver_id]] = 1
+    end
+  end
+  return driver_ride_count
+end
+
+p "Number of rides completed per driver:"
+p num_rides(ride_share_data)
+
+
 # - the total amount of money each driver has made
+def driver_earnings(structured_data)
+  earnings = {}
+  structured_data[:rides].each do |ride|
+    if earnings.include?(ride[:driver_id])
+      earnings[ride[:driver_id]] += ride[:cost]
+    else
+      earnings[ride[:driver_id]] = ride[:cost]
+    end
+  end
+  return earnings
+end
+
+p "Total cost per driver:"
+p driver_earnings(ride_share_data)
+
+
 # - the average rating for each driver
+def average_rating(structured_data)
+  avg_driver_ratings = {}
+
+  structured_data[:rides].each do |ride|
+    if avg_driver_ratings.include?(ride[:driver_id])
+      avg_driver_ratings[ride[:driver_id]] << ride[:rating].to_f
+    else
+      avg_driver_ratings[ride[:driver_id]] = [ride[:rating].to_f]
+    end
+  end
+
+  avg_driver_ratings.each do |driver_id, ratings|
+    avg_driver_ratings[driver_id] = (ratings.sum / ratings.size).round(1)
+  end
+
+  return avg_driver_ratings
+end
+
+p "Average rating per driver:"
+p average_rating(ride_share_data)
+
+
 # - Which driver made the most money?
-# - Which driver has the highest average rating?
\ No newline at end of file
+def max_earner(structured_data)
+  earnings = driver_earnings(structured_data)
+  max_earnings = earnings.values.max
+
+  if earnings.values.count(max_earnings) > 1
+    max_earner = (earnings.map { |driver, earnings| driver if earnings == max_earnings }).compact
+  else
+    max_earner = earnings.key(max_earnings)
+  end
+
+  return max_earner
+end
+
+p "Max earning driver(s):"
+p max_earner(ride_share_data)
+
+
+# - Which driver has the highest average rating?
+def highest_rated(structured_data)
+  ratings = average_rating(structured_data)
+  highest_rating = ratings.values.max
+
+  if ratings.values.count(highest_rating) > 1
+    highest_rated_driver = (ratings.map { |driver, avg_rating| driver if avg_rating == highest_rating }).compact
+  else
+    highest_rated_driver = ratings.key(highest_rating)
+  end
+
+  return highest_rated_driver
+end
+
+p "Highest rated driver(s):"
+p highest_rated(ride_share_data)
```

Comment:
Consider moving all code focused on user output to the bottom of your code so the program is organized where all methods definitions are grouped together, and all code focused on output is grouped together.

Reply:
Also, you made great choices about how to organize your data in hashes. Consider writing some driver code so the output is nicely formatted for the user -- you could even use `awesome_print`!